### PR TITLE
bpo-39171: Missing root in tkinter simpledialog.py _QueryDialog

### PR DESCRIPTION
**Pull request by Dominic-Mayers**
My first pull request. I hope this will be obviously linked to https://bugs.python.org/issue39171. 

https://bugs.python.org/issue39171

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@MaharishiCanada

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by MaharishiCanada** [comment]
I signed the CLA and added a blurb. I hope these are OK now.

File: Lib/tkinter/simpledialog.py [hunk 1]
```diff
@@ -260,7 +260,15 @@ def __init__(self, title, prompt,
                  parent = None):
 
         if not parent:
-            parent = tkinter._default_root
+            if tkinter._default_root:
```
Comment (E-Paine) [inline Lib/tkinter/simpledialog.py:263]: I would recommend changing this to an explicit check for "not None"

File: Lib/tkinter/simpledialog.py [hunk 2]
```diff
@@ -260,7 +260,15 @@ def __init__(self, title, prompt,
...
+                parent = tkinter._default_root
+            elif tkinter._support_default_root:
+                parent = tkinter.Tk()
+                parent.withdraw()
```
Comment (E-Paine) [inline Lib/tkinter/simpledialog.py:267]: If a new Tk instance is being created, I am of the opinion it should be explicitly deleted once we are done with the dialog (we shouldn't rely on the GC to clean it).

File: Misc/NEWS.d/next/Core and Builtins/2019-12-31-19-06-02.bpo-39171.cmXDsX.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Creating missing parent in tkinter simpledialog when possible.
```
Comment (E-Paine) [inline Misc/NEWS.d/next/Core and Builtins/2019-12-31-19-06-02.bpo-39171.cmXDsX.rst:1]: Would this be better suited to the "library" part of the news?

File: Lib/tkinter/simpledialog.py (hunk 2 above)
Comment (E-Paine) [reply Lib/tkinter/simpledialog.py:267]: On second thought, would it be better to mimic tkinter.commondialog (for consistency), which doesn't withdraw or delete a new window?

File: Lib/tkinter/simpledialog.py (hunk 1 above)
Comment (Dominic-Mayers) [reply Lib/tkinter/simpledialog.py:263]: Yes. The only reason why I did not do it, is to make as little change as possible.

File: Lib/tkinter/simpledialog.py (hunk 2 above)
Comment (Dominic-Mayers) [reply Lib/tkinter/simpledialog.py:267]: I thought about that too. It was a few months ago, but I remember to have compared with what is done elsewhere and came to the same conclusion.

File: Misc/NEWS.d/next/Core and Builtins/2019-12-31-19-06-02.bpo-39171.cmXDsX.rst (hunk 3 above)
Comment (Dominic-Mayers) [reply Misc/NEWS.d/next/Core and Builtins/2019-12-31-19-06-02.bpo-39171.cmXDsX.rst:1]: I don't know much about these aspects.

**Comment by E-Paine** [comment]
I really appreciate the effort that has clearly gone into this patch, however I fear it is needlessly over-complicated. It may be better to set the parent to a dummy frame, similar to 'commondialog', which will then handle all your logic in a single line (``parent = Frame(parent)`` - though you would want to rename one of those "parent"s for clarity).

**Comment by MaharishiCanada** [comment]
It is not clear to me that it will be equivalent. I suspect that to have the expected behavior with regard to the location of the widget, one may want to reuse the current main window, if it exists. Honestly, the current code is something that I wrote few months ago, so I am relying on vague memory, but what I recall is that location with respect to the current parent was important.  But, if you are sure that you are taking care of all the logic with respect to location, go ahead.

**Comment by E-Paine** [comment]
Sorry, by a dummy widget, I meant create a blank frame but not add it to the window. This way the window supplied by "parent" is used if it exists, otherwise the default root and falling back to create a new window. The widget is never added to the window using a display manager.
Unfortunately I cannot make changes to your branch as I am not a core-dev, but the changes would only be to remove all ``if not parent:`` code (including the if statement) and replace it with ``parent = Frame(parent)``.
As for the change to the news entry, the file could just be moved to Misc/NEWS.d/next/Library/2019-12-31-19-06-02.bpo-39171.cmXDsX.rst with no changes required to the content ("Core and Builtins" is generally for changes to the Python interpreter)

**Comment by MaharishiCanada** [comment]
I propose this for the news: `Creating missing parent in tkinter simpledialog when possible. Patch by E-Paine and MaharishiCanada.` I will use `parent = Frame(parent)` as you suggest for the revised patch. It makes plenty of sense that this will reuse the parent if it exists. I'll also check the behavior with the use case I have in mind.

**Comment by serhiy-storchaka** [comment]
It was done in more general way in other issues.

**Comment by MaharishiCanada** [comment]
Great !

**Pull request closed without merging**